### PR TITLE
Handle Conversions from Dse/Cassandra data type to Avro

### PR DESCRIPTION
**Pull request by tomekl007**
TODO: 
- [x] handle tuple
- [x] handle UDT
- [x] handle custom type
- [x] more unit tests

### CODE REVIEW

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  1
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  1
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

File: producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java [hunk 1]
```diff
@@ -15,24 +15,200 @@
...
+      return createCustomSchema((Custom) type);
+    } else if (type instanceof Native) {
+      return createNativeSchema((Native) type);
     } else {
-      // todo handle other types
       throw new UnsupportedOperationException(String.format("The type: %s is not supported", type));
     }
   }
+
+  private static Schema createNativeSchema(Native type) {
```
Comment (jorgebay) [inline producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java:120]: NIT: change the name to `getNativeSchema`

File: producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    return (bytes == null) || bytes.remaining() == 0 ? null : new BigInteger(Bytes.getArray(bytes));
+  }
+
+  public ByteBuffer toBytes(BigInteger value, Schema schema, LogicalType type) {
+    return (value == null) ? null : ByteBuffer.wrap(value.toByteArray());
+  }
+
+  @Override
+  public Schema getRecommendedSchema() {
+    return Schema.create(Type.INT);
```
Comment (jorgebay) [inline producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java:47]: Shouldn't this be `Type.LONG` or are all ints in avro vints?

**Review by jorgebay — COMMENTED** [review]
I took a first pass at it and its looking great, I've added few comments below.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- producer-kafka/src/test/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverterTest.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
- producer-kafka/src/test/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverterTest.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/test/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverterTest.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/test/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverterTest.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

File: producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java (hunk 2 above)
Comment (tomekl007) [reply producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java:47]: Actually, it follows the same pattern as `DecimalConversion` that is provided by Avro, and the recommended type should be `Type.Bytes`. I've changed it - good catch

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- producer-kafka/src/test/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverterTest.java  1
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteConversion.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/KeyValueConstructor.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/CqlToAvroTypeConverter.java  4
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  2
- producer-kafka/src/test/java/io/stargate/producer/kafka/helpers/MutationEventHelper.java  2
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/BigIntegerLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ByteLogicalType.java  3
- producer-kafka/src/main/java/io/stargate/producer/kafka/schema/codecs/ShortConversion.java  3
         

Clones added
============
- producer-kafka/src/test/java/io/stargate/producer/kafka/KafkaCDCProducerIntegrationTest.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6311317&bid=20521764)

**Pull request merged**